File: probs/121.cpp
#include <vector>
#include <iostream>
#include <string>
using namespace std;

int maxProfit(vector<int> &prices)
{
    vector<int> best(prices.size(), 1000000000);

    int result = 0;
    for (int i = 1; i < best.size(); i++)
    {
        best[i] = min(prices[i - 1], best[i - 1]);

        result = max(result, prices[i] - best[i]);
    }

    return result;
}

File: zig/121.zig
const std = @import("std");

fn solution(nums: []i32) !i32 {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();

    var best: i32 = 0;
    var lows = std.ArrayList(i32).init(allocator);
    defer lows.deinit();

    try lows.resize(nums.len);
    lows.items[0] = nums[0];

    for (1..nums.len) |idx| {
        lows.items[idx] = @min(nums[idx], lows.items[idx - 1]);

        best = @max(best, nums[idx] - lows.items[idx]);
    }

    return best;
}

const expect = std.testing.expect;

test "test case #1" {
    var nums = [_]i32{ 7, 1, 5, 3, 6, 4 };

    const profit = try solution(&nums);

    try expect(profit == 5);
}

test "test case #2" {
    var nums = [_]i32{ 7, 6, 4, 3, 1 };

    const profit = try solution(&nums);

    try expect(profit == 0);
}
